features/253-auth-sdk-callback-in-next/api.md

```diff
@@ -0,0 +1,215 @@
+```
+
+In skygear next, we will have several gears, and each gear handles its own responsibility only. For user auth related functionalities, we will use two gears to handle a user's state.
+
+- auth gear
+    - it handles auth related information:
+        - user role, disabled state, verified state, etc.
+        - principal (how a user can login, e.g. (username/email) + password)
+- record gear (optional).
+    - a general purpose gear to handle user profile.
+    - accessible by record gear only.
+    - expected to be accessed by Auth gear through public API optionally.
+
+This feature PR aims to find out a graceful design of the data type of the auth gear response object.
+
+# Glossary
+
+- skygear next: next generation skygear.
+- gear: distinct functionality component of skygear next.
+- auth gear: skygear next component for user authentication and authorization.
+- record gear: general purpose skygear next component handles record operation (CRUD).
+- profile record: a record object which handles user profile information (phone number, address, ...).
+- auth info: user authentication related information, like role, disabled state, verified state, principal.
+
+
+# Scenario
+
+- [user] sign up and login
+- [user] access auth info (roles, disabled)
+- [user] access user profile (username, email, phone number, ...)
+- [user] update user profile (username, email, phone number, ...)
+- [admin] query other users by some predicates
+- [admin] set other users auth info (roles, disabled)
+
+# Guideline
+
+For a developer, it's not easy to understand different data types convey similar concepts. Before skygear next, we had already found `authUser` and `UserRecord` confuse developers, both of them are considered as `user` data type and misuse them easily on SDK.
+
+So in skygear next, even though record gear is optional, we won't introduce a new `coreUser` data type, on the contrary, we will keep return data type of auth gear as `record` to reduce any possible misunderstanding. 
+
+For the case record gear is missing, auth gear will have a simple user profile implementatio by adding a json field to `_core_user` table. And it will also generate a record alike data arribute to client SDK.
+
+# Example usage on JS SDK
+
+keep current design, return data type is `record`.
+      
+```
++--------------------------+
+|                          |
+| [ currentUser <record> ] |
+| username                 |
+| email                    |
+| phone                    |
+| .....                    |
+|                          |
++--------------------------+
+
+// user record
+skygear.auth.currentUser;
+
+// current user auth info
+skygear.auth.isCurrentUserVerified;
+skygear.auth.isCurrentUserDisabled;
+
+// current user profile record
+const phone = skygear.auth.currentUser['phone'];
+
+// update record
+skygear.auth.currentUser['username'] = 'new-username';
+skygear.auth.currentUser['phone'] = 'new-phone';
+skygear.auth.updateUser(skygear.auth.currentUser);
```

Comment:
@chpapa I am still writing down some example codes about how to access user auth info, before I finish, let me add some short responses to you. 

Followings are based on current implementation direction,:

1. for `skygear.db.save(user)` case, this behavior is forbidden, a user can only update user profile via new `skygear.auth.updateUser(...)` API through auth gear to record gear. The reason here is for auth data (username, email) consistency.
2. for other user auth properties (roles, last login at, last seen at, ...),
   - for current user, they can be accessed via new properties, like: `skygear.auth.currentUserLastLoginAt`, `skygear.auth.currentUserSignupAt`.
   - for the case to access other user's auth properties, they can be accessed via new APIs, like: `skygear.auth.getUserLastLoginAt(userRecord)`, `skygear.auth.getUserSignupAt(userRecord)`

Feel free to give me your concerns. Thanks.

Reply:
Thanks @kaojohnny !

1. Okay, so we need to think about how to explain this to users...
2. Shall we consider generalize how we treat current users and other users? E.g. change `skygear.auth.currentUserLastLoginAt` to `skygear.auth.getUserLastLoginAt(skygear.auth.currentUser)`?
    * Main reason is just thought it would help make the API references / guide easier? (Is it easier to understand?)

Reply:
Thanks for your comments @chpapa, I will add them to another issue (https://github.com/SkygearIO/features/issues/256).